Added app.get("/user/:handle/books") Web api

MartinHwang opened the pull request:
app.get("/user/:handle/books", getUserBookDetails);
Modified books.js to upload book image but still need fixing

MartinHwang committed:
Added app.get("/user/:handle/books") Web api
app.get("/user/:handle/books", getUserBookDetails);
Modified books.js to upload book image but still need fixing

Harukaichii commented on functions/handlers/books.js:
```diff
@@ -17,7 +17,7 @@ exports.getAllBooks = (req, res) => {
           commentCount: doc.data().commentCount,
           favCount: doc.data().favCount,
           chapterCount: doc.data().chapterCount,
-          userImage: doc.data().userImage,
```
Hold up. We still need 

`userImage : doc.data().userImage `

If I'm not mistaken, when we display the book infromation, we still need to display the user's profile picture right? if we take it out here we probably can't do that !

MartinHwang replied:
In my opinion, we don't need to display user's profile picture because books which belong to user are displayed in the content creation page, /publish/username. What do you think about it?

Harukaichii replied:
Yes but if you're a user and you're looking at someone else's book, how are you going to display that person's picture?

MartinHwang replied:
> Yes but if you're a user and you're looking at someone else's book, how are you going to display that person's picture?
Oh, I see. we need book of image and person's picture.

Harukaichii commented on functions/handlers/books.js:
```diff
@@ -52,6 +53,7 @@ exports.postOneBook = (req, res) => {
     })
     .catch((err) => {
       res.status(500).json({ error: "Something went wrong :(" });
+      res.json(err);
```
we already have  
`res.status(500).json({ error: "Something went wrong :(" });` 

so the response given will be the object { error: "Something went wrong :(" } not err. If you want to return the error object instead change the above line to 
` res.status(500).json(err)`

MartinHwang replied:
For sure, I will change it.

Harukaichii commented on functions/handlers/books.js:
```diff
@@ -65,7 +67,7 @@ exports.editBook = (req, res) => {
     desc: req.body.desc,
     title: req.body.title,
     userHandle: req.user.handle,
-    imageUrl: req.user.imageUrl,
```
I think you should still leave the userImage in here just in case we need to display their picture on the edit page

MartinHwang replied:
I did not complete to uploading book image yet. I am still working with this :)

Harukaichii replied:
Alrighty : )

Harukaichii commented on functions/handlers/books.js:
```diff
@@ -127,7 +129,7 @@ exports.commentOnBook = (req, res) => {
     createdAt: new Date().toISOString(),
     bookId: req.params.bookId,
     userHandle: req.user.handle,
-    userImage: req.user.imageUrl,
```
same comment as above. When commenting on a book if we don't carry over userImage, we can't display the userImage whene we're rending in frontend

Harukaichii commented on functions/handlers/users.js:
```diff
@@ -149,6 +149,46 @@ exports.getUserDetails = (req, res) => {
+    .then(data => {
+      userData.books = [];
+      data.forEach(doc => {
+        userData.books.push({
```
I think you're missing the field bookImage here :)

Harukaichii commented on functions/handlers/books.js:
```diff
@@ -264,3 +266,57 @@ exports.deleteBook = (req, res) => {
+      })
+      .then(() => {
+        const bookImage = `https://firebasestorage.googleapis.com/v0/b/${config.storageBucket}/o/${imageFileName}?alt=media`;
+        return db.doc(`/users/${req.user.handle}`).update({ bookImage });
```
I think this should be a url to a "/book/${req.params.bookId}" or something similar

definitely not to users.

MartinHwang replied:
Yes, I will use it for upload book image module

Harukaichii commented on functions/handlers/books.js:
```diff
@@ -264,3 +266,57 @@ exports.deleteBook = (req, res) => {
+            contentType: imageToBeUploaded.mimetype
+          }
+        }
+      })
```
from line 271 - 300 , I think you can make a utility function out of it. The real difference between uploading userImage and bookImage is the response.

If you want to be fancy, you can turn uploadBookImage and uploadUserImage into 1 function. The response  can then change depending on a variable you pass when you make the request

MartinHwang replied:
Sounds good. I will consider it. Thanks

Harukaichii reviewed (changes requested):
There's somethings that you should probably checkover. Especially the upload bookImage function. I tried to test it on Postman but I can't seem to upload an image after trying to post to /book/bookImage

MartinHwang committed:
Wroking with books.js and index.js

Book.js

/ Change book image
exports.onBookImageChange = functions
.region("us-central1")
.firestore.document("/books/{bookId}")
.onUpdate((change) => {
  console.log(change.before.data());
  console.log(change.after.data());
  if (change.before.data().bookImage !== change.after.data().bookImage) {
    console.log("Book image has changed");
    const batch = db.batch();
    return db
      .collection("books")
      .where("userHandle", "==", change.before.data().handle)
      .get()
      .then((data) => {
        data.forEach((doc) => {
          const book = db.doc(`/books/${doc.id}`);
          batch.update(book, { bookImage: change.after.data().bookImage });
        });
        return batch.commit();
      });
  } else return true;
});

Books.js:
//post a book
exports.postOneBook = (req, res) => {
  const { errors, valid } = validateFormat(req.body, ["title", "desc"]);
  const noBookImg = "no-book-img.png";
  if (!valid) return res.status(400).json(errors);
  const newBook = {
    title: req.body.title,
    desc: req.body.desc,
    userHandle: req.user.handle,
    userImage: req.user.imageUrl,
    bookImageUrl: `https://firebasestorage.googleapis.com/v0/b/${config.storageBucket}/o/${noBookImg}?alt=media`,
    createdAt: new Date().toISOString(),
    favCount: 0,
    commentCount: 0,
    chapterCount: 0,
  };

  db.collection("books")
    .add(newBook)
    .then((doc) => {
      const resBook = newBook;
      resBook.bookId = doc.id;
      res.json(resBook);
    })
    .catch((err) => {
      res.status(500).json({ error: "Something went wrong :(" });
      res.json(err);
      console.error(err);
    });
};

Commit:
adjusted backend to support uploading bookImage better

MartinHwang committed:
Renamed uploadBookImage and generated initialImage

PT-17
https://proudtale.atlassian.net/browse/PT-17?atlOrigin=eyJpIjoiMzU0NDJmYmI1YTgyNDkwZmFhNjk3M2FkZjY1MTBmZmQiLCJwIjoiaiJ9

PT-18
https://proudtale.atlassian.net/browse/PT-18?atlOrigin=eyJpIjoiZmIyM2Q5MGUxNjEzNDRhNDgwZGM0YjU5MmNjNDIwZjYiLCJwIjoiaiJ9

PT-19
https://proudtale.atlassian.net/browse/PT-19?atlOrigin=eyJpIjoiYmVhMDcwNzM5OTA3NDMyMmE3YzQ3YmNkZjM0ZDgyNzkiLCJwIjoiaiJ9

PT-51
https://proudtale.atlassian.net/browse/PT-51?atlOrigin=eyJpIjoiMDZjNGE5Y2M1YWVmNGM5NTljMWEyNmNkNzhiZmIwNGEiLCJwIjoiaiJ9

Implemeted codes which are associated with PT-17, PT-18, PT-19, PT-51

mdc2tanu committed:
Merge branch 'bookImage/develop-MartinHwang-PT-18-53' into bookImage/Queenie/Mod

mdc2tanu committed:
Merge pull request #5 from proudtale/bookImage/Queenie/Mod

adjusted backend to support uploading bookImage better
